File: backend/app/models/schedule.py
from __future__ import annotations

from datetime import date, datetime

from sqlalchemy import Date, DateTime, ForeignKey, Integer, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db import Base

CYCLE_STATUSES = ("planned", "running", "completed", "aborted")
CELL_USE_STATUSES = ("planned", "started", "completed", "failed", "aborted", "cancelled")


class RunBatch(Base):
    """A Run: one SMRT Link run design = one physical load session on one instrument on
    one day. Holds 1-2 Plates (its ``cycles``) that together load up to 8 cells across up
    to two sample plates. Uniquely identified by (instrument, load_date).

    Maps onto the PacBio/SMRT Link "run design" (Plate 1 + Plate 2); see
    docs/pacbio-sprq-nx-scheduling-reference.md. ``run_name`` (e.g. Sanger's
    "TRACTION-RUN-1234") is the run-level Traction ID. Historically this was one
    instrument-day with a single 1:1 Cycle; the Run->Plate split promoted RunBatch to the
    run and Cycle to a plate."""

    __tablename__ = "run_batches"
    __table_args__ = (UniqueConstraint("instrument_id", "load_date", name="uq_run_batch_instrument_load_date"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    instrument_id: Mapped[int] = mapped_column(ForeignKey("instruments.id"), index=True)
    # The day the whole run is physically loaded on the instrument (one session). Plate 1
    # acquires on this day; a reuse Plate 2 acquires later but is still loaded now.
    load_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    # Lab-assigned run label (Traction ID), run-level, set at Confirm loaded. Moved here
    # from Cycle in the Run->Plate split - a run has one Traction ID across both plates.
    run_name: Mapped[str | None] = mapped_column(String(128), nullable=True)

    instrument: Mapped["Instrument"] = relationship(back_populates="run_batches")
    cycles: Mapped[list["Cycle"]] = relationship(
        back_populates="run_batch", cascade="all, delete-orphan", order_by="Cycle.plate_index"
    )


class Cycle(Base):
    """A Plate: one acquisition round within a Run (its parent RunBatch). A run has 1-2
    plates (``plate_index`` 1 or 2). Plate 1 acquires on the run's load_date; Plate 2
    acquires the same day when it is a second fresh tray (parallel - the single-use 8-cell
    run) or a later day when it reuses Plate 1's cells (sequential, after the on-board wash).
    Each plate holds up to 4 wells (its ``cell_uses``).

    Run time is per-well (CellUse.run_time_hours) - a deliberate divergence from the vendor's
    "one acquisition time per plate" default; see docs/pacbio-sprq-nx-scheduling-reference.md.
    ``movie_hours`` here is this plate's *representative* (longest) run time, kept in sync by
    placement_service.recompute_cycle_timing; it drives planned_end_at and the instrument lock.
    (Historically this class was 1:1 with RunBatch and modelled the whole run; the Run->Plate
    split promoted RunBatch to the run and this to a plate.)"""

    __tablename__ = "cycles"
    __table_args__ = (UniqueConstraint("run_batch_id", "plate_index", name="uq_cycle_run_plate"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    run_batch_id: Mapped[int] = mapped_column(ForeignKey("run_batches.id", ondelete="CASCADE"), index=True)
    # 1 or 2 - which sample plate / loading position this acquisition round is. server_default
    # is for the migration's existing rows only; every code path sets it explicitly.
    plate_index: Mapped[int] = mapped_column(Integer, nullable=False, server_default="1")
    # The day THIS plate sequences. == run.load_date for Plate 1 (and a same-day parallel
    # Plate 2); a later date for a reuse Plate 2. The chronological anchor for Use 1/2/3.
    acquire_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    # Representative (= longest) run time across this plate's cell_uses; see class docstring.
    movie_hours: Mapped[int] = mapped_column(Integer)
    planned_start_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    planned_end_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    actual_start_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    actual_end_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="planned", index=True)

    run_batch: Mapped["RunBatch"] = relationship(back_populates="cycles")
    cell_uses: Mapped[list["CellUse"]] = relationship(
        back_populates="cycle", cascade="all, delete-orphan", order_by="CellUse.well"
    )


class CellUse(Base):
    """One sample loaded on one cell in one well of a plate (its ``cycle``) - the "stage".
    Since ``cycle`` is now a Plate, UniqueConstraint(cycle_id, well) is per-plate: a reused
    cell legitimately appears in the same well (e.g. A01) on both plates of one run - those
    are two different cycle_ids, so they never collide."""

    __tablename__ = "cell_uses"
    __table_args__ = (UniqueConstraint("cycle_id", "well", name="uq_cell_use_cycle_well"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    cycle_id: Mapped[int] = mapped_column(ForeignKey("cycles.id", ondelete="CASCADE"), index=True)
    cell_id: Mapped[int] = mapped_column(ForeignKey("cells.id"), index=True)
    sample_id: Mapped[int | None] = mapped_column(ForeignKey("samples.id"), nullable=True, index=True)
    well: Mapped[str] = mapped_column(String(8))
    # This well's own movie / run time (12/24/30 h). Set from the Run Design "Movie / run
    # time" dial when the sample is placed or auto-scheduled, then editable per-cell from the
    # slot-detail popover. Different wells of the same run may carry different values - the
    # owning Cycle.movie_hours tracks the longest of them (see Cycle docstring). server_default
    # is for the migration's existing rows only; every code path sets it explicitly.
    run_time_hours: Mapped[int] = mapped_column(Integer, nullable=False, server_default="24")
    status: Mapped[str] = mapped_column(String(20), default="planned", index=True)
    # Set by a Cell QC tray re-zip (services/qc_service.py) when a failed/retired cell's
    # loss shifts this acquisition onto a DIFFERENT physical cell than originally planned:
    # holds the cell_id it was planned on. Null = ran on the planned cell. Drives the grid's
    # "reassigned" flag and lets undo restore the original cell.
    reassigned_from_cell_id: Mapped[int | None] = mapped_column(
        ForeignKey("cells.id"), nullable=True
    )
    outcome_notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    # Free-text note the lab user attaches to this sample-on-this-cell placement. Distinct
    # from outcome_notes (which records a QC verdict via Mark Failed / Stop cell): this is a
    # general annotation, editable at any time - including after the run is locked - and
    # surfaced on the slot-detail popover and the printed batch sheet.
    notes: Mapped[str | None] = mapped_column(Text, nullable=True)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

    cycle: Mapped["Cycle"] = relationship(back_populates="cell_uses")
    # foreign_keys pins this to cell_id: there are now two FKs to cells (cell_id and the QC
    # reassigned_from_cell_id), so the relationship join must be disambiguated.
    cell: Mapped["Cell"] = relationship(back_populates="cell_uses", foreign_keys=[cell_id])
    sample: Mapped["Sample | None"] = relationship(back_populates="cell_uses")
    barcodes: Mapped[list["CellUseBarcode"]] = relationship(back_populates="cell_use", cascade="all, delete-orphan")

    @property
    def barcode_list(self) -> list[str]:
        return [b.barcode for b in self.barcodes]


class CellUseBarcode(Base):
    """Barcode snapshot per cell use - not a live join to Sample, so history stays correct
    even if a sample record is later corrected."""

    __tablename__ = "cell_use_barcodes"

    id: Mapped[int] = mapped_column(primary_key=True)
    cell_use_id: Mapped[int] = mapped_column(ForeignKey("cell_uses.id", ondelete="CASCADE"), index=True)
    barcode: Mapped[str] = mapped_column(String(64), index=True)

    cell_use: Mapped["CellUse"] = relationship(back_populates="barcodes")

File: frontend/src/types/schedule.ts
/** Backend-schema mirror for the run-oriented scheduler (Run + Plate + Stage). Field names
 * are the wire format (no camelCase transform layer), so this file is the source of
 * truth for the frontend/backend contract - keep it in lockstep with backend/app/schemas/
 * run.py. A "run" is now a SMRT Link run design: one physical load session on one
 * instrument (RunOut, keyed by load_date), holding 1-2 plates (PlateOut, each an
 * acquisition round with its own acquire_date). A reuse run reads as one run whose Plate 2
 * acquires a later day. */

import type { CycleStatus } from "./common";

export type MaxUses = 1 | 2 | 3;
export type RunTimeHours = 12 | 24 | 30;
/** The auto-schedule strategies the UI surfaces, stored as the engine's own mode names so
 * they pass straight through to /api/auto-fill (which still accepts the wider legacy set -
 * see backend/app/schemas/run.py). The UI labels them "Fastest", "Efficient" and "By order"
 * (see RunDesignFields.OBJECTIVE_OPTIONS):
 *   - "utilisation" ("Fastest"): open enough distinct fresh cells to fill the tray so
 *     every sample starts as soon as possible - each cell then carries an expiry timer.
 *   - "fewest" ("Efficient"): reuse a cell to its Max-uses depth before opening the next,
 *     so fewer cells have a running 108h window at once.
 *   - "order" ("By order"): place samples strictly in the sequence they were uploaded, and
 *     within each upload the order their rows appeared in the CSV (fills the grid day-by-day
 *     in that sequence; ignores priority). Uses the same fill-a-tray cell choice as "Fastest". */
export type Objective = "fewest" | "utilisation" | "order";
/** Per-day cap on how many samples auto-fill may schedule on one instrument: 4 = one sample
 * plate, 8 = two. A cap on samples, NOT a cell-tray count - the samples run on cells drawn from
 * whatever tray(s) are loaded (either carousel bay). Surfaced in the UI as "Plates per run". */
export type CellsPerDay = 4 | 8;
/** Grid POSITION within a run, 0-7: Plate 1 -> 0-3, Plate 2 -> 4-7. NOT the physical well
 * index - a reuse Plate 2 sits in the same wells (A01-D01) as Plate 1, so its stages carry
 * slot_index 4-7 while their `well` repeats A01-D01. */
export type SlotIndex = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7;

export interface StageOut {
  /** Grid position 0-7: Plate 1 -> 0-3, Plate 2 -> 4-7 (see SlotIndex). */
  slot_index: SlotIndex;
  /** The plate LOADING position this sample was dropped onto (drives slot_index). A grid slot
   * is a loading position, not a cell - which physical cell runs here is separate (cell_id /
   * cell_home_well). For a reuse Plate 2 this can repeat A01-D01 even though slot_index is 4-7. */
  well: string;
  cell_use_id: number;
  cell_id: number;
  cell_ref: string; // Cell.code
  /** The physical cell's own tray identity well (A01-D01 / A02-D02) - its fixed A/B/C/D tray
   * position, distinct from `well` (the loading slot). Drives the ticket-stub letter (e.g.
   * "B1" = tray position B, Use 1) so the stub names the real cell even when the sample sits in
   * a different plate slot. Null for a legacy/bootstrap cell with no tray. */
  cell_home_well: string | null;
  /** 1-based position of this cell_use among its cell's loads - drives the Use 1/2/3 colour. */
  use_number: number;
  /** The physical cell's own use cap (usually 3; lower only if QC reduced it). With use_number
   * this gives the cell's remaining uses (cell_max_uses - use_number) shown on the Revio-screen
   * panel's "Remaining SMRT Cell uses" boxes. */
  cell_max_uses: number;
  /** This well's own movie / run time (h). Per-cell: set from the Run Design dial on
   * placement/auto-schedule, editable per-cell in the slot-detail popover. Different wells of
   * one run may differ; PlateOut.movie_hours is the longest of its plate's wells. */
  run_time_hours: RunTimeHours;
  sample_id: number | null;
  sample_pool_id: string | null;
  /** Library insert / fragment size (bp) of the sample in this slot, or null if not recorded.
   * Drives the grid card's "[<5kb]" flag and the small-insert-on-reuse warning (a small-insert
   * sample sitting on use_number >= 2). The threshold is admin-configurable (read client-side). */
  insert_size_bp: number | null;
  /** Duplicate marker for the sample in this slot: when its Pool ID is carried by more
   * than one sample (any status), duplicate_total is the count and duplicate_index this copy's
   * 1-based position. Both null/absent for a one-off — the grid card shows "1/3" only when set. */
  duplicate_index?: number | null;
  duplicate_total?: number | null;
  /** True when this cell was already used by another copy of the exact same Pool ID (a
   * sibling duplicate sharing a barcode) - an intentionally ALLOWED reuse, not a clash: reusing
   * a cell with the identical sample can't misattribute reads to a foreign sample, so there's
   * no cross-contamination risk the barcode-clash rule exists to prevent. Shown so it's
   * transparent at a glance rather than a silent exception. See docs/pacbio-sprq-nx-scheduling-
   * reference.md's barcode-carryover rule. */
  duplicate_cell_reuse?: boolean;
  barcodes: string[];
  /** This specific use's own status (planned/started/completed/failed/cancelled). */
  cell_use_status: string;
  /** The physical cell's overall status (open/exhausted/window_expired/retired/stopped). */
  cell_status: string;
  /** True if any use of this cell has a recorded "failed" outcome - lets the grid tell an
   * earlier, still-untouched use apart from the one a Stop cell was actually triggered
   * from once the cell goes "stopped" (see SchedulerSlotView's qcAlert). */
  cell_has_failed_use: boolean;
  /** 1-4 position within this cell's physical SPRQ-Nx SMRT Cell tray - null for cells
   * with no tray (created before this feature, or via the bootstrap cutover tool). */
  tray_position: number | null;
  /** The physical tray this cell belongs to - lets the grid's per-tray "Discard Cells"
   * action target every sibling cell, not just the ones with a filled slot this cycle. */
  tray_id: number | null;
  /** Hours elapsed since this cell's own first use (null if not started yet) - drives the
   * slot's expiry shading. Per-cell, not per-tray - see docs/pacbio-sprq-nx-scheduling-
   * reference.md #2 (no shared tray-level clock). */
  window_hours_elapsed: number | null;
  /** True when this is a REUSE (Use 2/3) whose plate's planned start has slipped past the cell's
   * 108h reuse deadline - it can no longer physically start in time. Drives the card's "⚠ Window"
   * flag and the slot/cell popover's out-of-window Note + "Load fresh tray" action. False for a
   * first use and for any in-window reuse. Estimated until Use 1 is confirmed. */
  reuse_window_exceeded?: boolean;
  /** Advisory only, never blocks a placement - hours by which this use's own start preceded
   * its cell's real physical readiness (the prior use's movie end + the on-board reuse wash).
   * null when this is the cell's first use, or the start was already safely at/after
   * readiness. A distinct clock from window_hours_elapsed's 108h lifetime check. Only
   * populated on placement/move/auto-fill responses, null on the plain grid feed. */
  reuse_not_ready_hours: number | null;
  /** Free-text note the user attached to this sample-on-this-cell placement, shown and
   * editable in the slot-detail popover. Distinct from the QC outcome note. */
  notes: string | null;
  /** Cell QC reconciliation: true if a QC tray re-zip shifted this acquisition onto a
   * different physical cell than planned (drives a "reassigned" grid marker). */
  reassigned?: boolean;
  /** True if that reassignment landed on a cell that already burned a clashing barcode. */
  barcode_clash?: boolean;
}

/** One acquisition round within a run (a persisted Cycle). Up to 4 wells. Plate 1 acquires
 * on the run's load_date; Plate 2 acquires the same day (fresh parallel second tray) or a
 * later weekday (reuse of Plate 1's cells). */
export interface PlateOut {
  plate_id: number; // the Cycle id
  plate_index: 1 | 2; // which sample plate / loading position
  acquire_date: string; // YYYY-MM-DD - the day THIS plate sequences (== run.load_date for Plate 1)
  /** True when this plate reuses an earlier plate's cells (same tray, a later acquire_date) -
   * its wells show Use >= 2. False for Plate 1 and for a fresh parallel Plate 2. */
  is_reuse: boolean;
  /** This plate's representative run time: the longest of its wells' per-cell run_time_hours. */
  movie_hours: number;
  status: CycleStatus;
  planned_start_at: string; // ISO datetime
  planned_end_at: string;
  actual_start_at: string | null;
  actual_end_at: string | null;
  /** ONLY filled wells - pad to slot_index 0-7 for rendering (see padStages). */
  stages: StageOut[];
}

export interface RunOut {
  run_id: number; // the RunBatch id
  instrument_serial: string;
  load_date: string; // YYYY-MM-DD - the day the whole run is physically loaded (one session)
  /** Optional lab-assigned label (e.g. "TRACTION-RUN-1234") set when the run is locked
   * (Confirm loaded) - overrides "#<run_id>" everywhere a run is displayed. */
  run_name: string | null;
  /** Derived run-level status: "running"/"completed" once its plates are, else "planned". */
  status: CycleStatus;
  /** The instrument is held until the last plate's acquisition finishes + buffer - when it
   * becomes available to *start* another run (loading the next run's cells is never blocked). */
  lock_until: string; // ISO datetime
  /** "now" falls within the run's load -> last-acquisition window and it isn't aborted/completed. */
  is_locked: boolean;
  /** Derived (never stored): when this run's cells actually break out once the instrument's OTHER
   * resident runs' sequencing-lane occupancy is accounted for. Only populated on placement / move /
   * auto-fill responses (null on the grid feed). If it's meaningfully later than the chosen load
   * time, the load was accepted but the cells queue — surfaced to the user as an advisory. */
  effective_start_at: string | null;
  /** True when effective_start_at is meaningfully later than the run's own load time. */
  starts_later_than_requested: boolean;
  /** 1-2 plates. Plate 1 first; a second plate is a fresh parallel tray or a reuse. */
  plates: PlateOut[];
}
